src/ray/raylet/node_manager.cc

```diff
@@ -847,18 +844,15 @@ void NodeManager::DispatchTasks(
         // once the first task is not feasible, we can break out of this loop
         break;
       }
-      if (AssignTask(task, &post_assign_callbacks)) {
-        removed_task_ids.insert(task_id);
+
+      // Try to get an idle worker to execute this task. If nullptr, there
+      // aren't any available workers so we can't assign the task.
+      std::shared_ptr<Worker> worker = worker_pool_.PopWorker(spec);
+      if (worker != nullptr) {
+        AssignTask(worker, task);
```

Comment:
Wasn't one of the issues here that AssignTask invalidates the iterator of this loop since it updates the queues?

Reply:
Good catch, reverted to using callbacks